src/services/refactors/convertReactPureComponent.ts

```diff
@@ -0,0 +1,351 @@
+                    const isStatic = val.modifiers && val.modifiers.some(mod => mod.kind === SyntaxKind.StaticKeyword);
+                    if (isMethodNamedRender && !isStatic) {
+                        render = val.body!;
+                        return true;
+                    }
+                }
+                // TODO: Should include staic `propTypes`, `contextTypes`, `defaultProps`, `displayName`. Not now.
+                return false;
+            });
+            // ? Now check reference to `state` or `setState`
+            if (convertable) {
+                // OK let's do this quick
+                if (removeEmpty(render!.getText(sourcefile)).match(/this\.(state|setState)/)) {
+                    convertable = false;
+                }
+            }
+            return { convertable, render: render! };
+        }
+        const icc = isConvertibleComponent(node, sourcefile);
+        if (!icc.convertable) return undefined;
+        return { name: ircs.name, propsType: ircs.propsType, render: icc.render, originNode: node };
+    }
+
+    function transformSFCtoComponent(component: Component, context: RefactorContext): RefactorEditInfo {
+        const changeTracker = textChanges.ChangeTracker.fromContext(context);
+
+        const react = getReferenceToReact(context.file)!;
+
+        const extendsClause = createHeritageClause(SyntaxKind.ExtendsKeyword, [
+            createExpressionWithTypeArguments(component.propsType ? [component.propsType] : undefined,
+                createIdentifier(react.PureComponent))
+        ]);
+        const renderMethod = createMethod([], [], void 0, "render",
+            void 0, void 0, [], void 0, component.render);
+
+        let newNode: ClassLikeDeclaration;
+        if (component.name) {
+            newNode = createClassDeclaration(void 0, void 0, component.name, void 0, [extendsClause], [renderMethod]);
+        }
+        else {
+            newNode = createClassExpression(void 0, void 0, void 0, [extendsClause], [renderMethod]);
+        }
+        changeTracker.replaceNode(context.file, component.originNode, newNode);
+        return {
+            edits: changeTracker.getChanges().map(x => ({
+                ...x, textChanges: x.textChanges.map(y => ({
+                    ...y, newText: y.newText.replace(/props/g, "this.props")
+                }))
+            }))
+        };
+    }
+
+    function transformComponentToSFC(component: Component, context: RefactorContext): RefactorEditInfo {
+        const changeTracker = textChanges.ChangeTracker.fromContext(context);
+
+        const react = getReferenceToReact(context.file)!;
+
+        const newNode: FunctionDeclaration = createFunctionDeclaration(
+            void 0, void 0, void 0, component.name, void 0,
+            [createParameter(void 0, void 0, void 0, "props", void 0)],
+            createTypeReferenceNode(react.SFC, component.propsType ? [component.propsType] : undefined),

```

Comment:
When I tested this, `react.SFC` was the string `"React.Component"`. That's not a valid identifier, so it causes assertion failures in the formatter (#24751). Should use an `EntityName` node instead.
Also, when a node in the old tree is moved to the new tree, you should use `getSynthesizedDeepClone` to get new nodes.

Reply:
Thanks, with your help I can continue working on it